src/Compiler/Microsoft.CodeAnalysis.Razor.Compiler/src/Language/Intermediate/IntermediateNodeFormatter.cs

```diff
@@ -1,17 +1,225 @@
+    // Some nodes have properties: Document
+    // Some nodes have both: TagHelperProperty
+    public enum FormatterMode
+    {
+        PreferContent,
+        PreferProperties,
+    }
+
+    private const string ShortNameSuffix = nameof(IntermediateNode);
+    private static readonly Dictionary<Type, string> s_nodeTypeToShortNameMap = [];
+
+    private static readonly char[] s_charsToEscape = ['\r', '\n', '\t'];
+
+    private readonly StringBuilder _builder = builder;
+    private readonly FormatterMode Mode = mode;
+    private readonly bool _includeSource = includeSource;
+
+    private readonly Dictionary<string, string> _properties = new(StringComparer.Ordinal);
+    private string? _content;
+
+    private static string GetNodeShortName(IntermediateNode node)
+    {
+        lock (s_nodeTypeToShortNameMap)
+        {
+            return s_nodeTypeToShortNameMap.GetOrAdd(node.GetType(), ComputeShortName);
+        }
+
+        static string ComputeShortName(Type type)
+        {
+            var shortName = type.Name;
+
+            if (shortName.EndsWith(ShortNameSuffix, StringComparison.Ordinal))
+            {
+                shortName = shortName[..^ShortNameSuffix.Length];
+            }
+
+            return shortName;
+        }
+    }
+
+    public void WriteChildren(IntermediateNodeCollection children)
+    {
+        Debug.Assert(children != null);
+
+        _builder.Append(" \"");
+
+        foreach (var child in children)
+        {
+            if (child is IntermediateToken token)
+            {
+                WriteEscaped(token.Content);
+            }
+        }
+
+        _builder.Append('"');
+    }
+
+    public void WriteContent(string? content)
+    {
+        if (content != null)
+        {
+            _content = content;
+        }
+    }
+
+    public void WriteProperty(string key, string? value)
+    {
+        Debug.Assert(key != null);
+
+        if (value != null)
+        {
+            _properties.Add(key, value);
+        }
+    }
+
+    public void FormatNode(IntermediateNode node)
+    {
+        _builder.Append(GetNodeShortName(node));
+
+        if (_includeSource)
+        {
+            _builder.Append(' ');
+            _builder.Append(node.Source?.ToString() ?? "(n/a)");
+        }
+
+        node.FormatNode(this);
+
+        if (ShouldWriteContent())
+        {
+            _builder.Append(" \"");
+            WriteEscaped(_content);
+            _builder.Append('"');
+        }
+
+        if (ShouldWriteProperties())
+        {
+            _builder.Append(" { ");
+
+            var first = true;
+
+            foreach (var (key, value) in _properties)
+            {
+                if (first)
+                {
+                    first = false;
+                }
+                else
+                {
+                    _builder.Append(", ");
+                }
+
+                _builder.Append(key);
+                _builder.Append(": \"");
+                WriteEscaped(value);
```

Comment:
>WriteEscaped(value); [](http://example.com/codeflow?start=16&length=20)

Intentionally handling escaped now?

Reply:
Prior to this change, property values were escaped when they were added to the dictionary. Now, they're escaped when written.